### PR TITLE
kvserver,rpc: set the stage for maintaining a local blocklist of permanently removed nodes

### PR DESCRIPTION
**Pull request by tbg**
As part of long-running migrations (#54903) we need to make sure, to the best of our abilities, that decommissioned nodes can not return to the cluster. We will achieve this by persisting this knowledge locally (in an eventually consistent way, via gossip) and ading checks in the RPC layer.

This PR stops short of adding the actual storage but sets the stage for doing so.

First seven commits are from another PR and should be ignored here.

### CODE REVIEW

**Review by knz — COMMENTED** [review]
I just reviewed r1, r2, r3,   r6,  r12, r13, r14. I'll let irfan review the rest.

In r13-r14 I recommend extending the commit message (and/or code comments) to remind the reader that Ping is used prior to actually using the RPC connection for inter-node traffic, so that its logic effectively gates the use of a connection. This wasn't clear at first.

> Reviewed 1 of 1 files at r1, 3 of 3 files at r2, 2 of 2 files at r3, 2 of 4 files at r4, 2 of 2 files at r6, 6 of 11 files at r7, 5 of 5 files at r12, 3 of 3 files at r13, 4 of 4 files at r14.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54936#-:-MIOqXuS3SAju0Aom_TX:b3tdlk1)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif, @knz, and @tbg)

___
*[pkg/cli/node.go, line 341 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/54936#-MIOpirr8E--_PanpIbD:-MIOpirr8E--_PanpIbE:b-llq3v7) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6ee769045607ed00afa4062e5cc0bb5a6f334ee/pkg/cli/node.go#L341)):*
> ```Go
> 	c := serverpb.NewAdminClient(conn)
> 	if err := runDecommissionNodeImpl(ctx, c, nodeCtx.nodeDecommissionWait, nodeIDs); err != nil {
> 		cause := errors.Cause(err)
> ```

`UnwrapAll` is clearer IMHO. `Cause` is only provided for compatibility with pkg/errors.
___
*[pkg/cli/node.go, line 583 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/54936#-MIOpyx86o0I4hg0NpvX:-MIOpyx86o0I4hg0NpvY:bsxkxj9) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6ee769045607ed00afa4062e5cc0bb5a6f334ee/pkg/cli/node.go#L583)):*
> ```Go
> 		// ValidateLivenessTransition in kvserverpb/liveness.go for where this
> 		// error is generated.
> 		if s, ok := status.FromError(err); ok && s.Code() == codes.FailedPrecondition {
> ```

I'd recommend extracting `cause := errors.UnwrapAll(err)` for consistency with thge other case.

Maybe even worth introducing a `FromError` wrapper in our own `grpcutil` package and have that unwrap in all cases.
___
*[pkg/kv/kvserver/node_liveness.go, line 278 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/54936#-MIOpEq96YYP6bRcJ6Vi:-MIOpEqACcAfl4s27K6K:b-dtx0zb) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef38ec677b880654b02c71c39a6dd4cb3795422d/pkg/kv/kvserver/node_liveness.go#L278)):*
> ```Go
> 	// best effort attempt at marking the node as draining. It can return
> 	// without an error after trying unsuccessfully a few times. Is that what we
> 	// want?
> ```

As the last person interested in this function, I didn't appreciate that DefaultRetryOptions had a finite timeout. With a retry-forever, this code should be fine?
If it does, then there's a bug I agree.

**Review by irfansharif — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54936)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif and @knz)

___
*[pkg/cli/node.go, line 341 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/54936#-MIOpirr8E--_PanpIbD:-MIQd3cA1cJxoBDrA-aM:b-bnjf9c) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6ee769045607ed00afa4062e5cc0bb5a6f334ee/pkg/cli/node.go#L341)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

`UnwrapAll` is clearer IMHO. `Cause` is only provided for compatibility with pkg/errors.
</blockquote></details>

Done in #54544. I keep getting lost in the very broad API net have in cockroachdb/errors because of these compatibility constraints, I need to sit down and go through all of it once for good.
___
*[pkg/cli/node.go, line 583 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/54936#-MIOpyx86o0I4hg0NpvX:-MIQdz6z5WXf4ZzL5Xo6:b7hyu5b) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6ee769045607ed00afa4062e5cc0bb5a6f334ee/pkg/cli/node.go#L583)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I'd recommend extracting `cause := errors.UnwrapAll(err)` for consistency with thge other case.

Maybe even worth introducing a `FromError` wrapper in our own `grpcutil` package and have that unwrap in all cases.
</blockquote></details>

Done in #54544. Ditto the comment above, lately I've found it faster to just throw some random error handling variant from cockroachdb/errors and relying on the reviewers to teach me where I'm wrong (so I'm clearly pretty confused by it all).
___
*[pkg/kv/kvserver/node_liveness.go, line 278 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/54936#-MIOpEq96YYP6bRcJ6Vi:-MIQcVSrD-a9KbsesUMq:b-aa3nbp) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef38ec677b880654b02c71c39a6dd4cb3795422d/pkg/kv/kvserver/node_liveness.go#L278)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

As the last person interested in this function, I didn't appreciate that DefaultRetryOptions had a finite timeout. With a retry-forever, this code should be fine?
If it does, then there's a bug I agree.
</blockquote></details>

Patched in #54544 (https://github.com/cockroachdb/cockroach/pull/54544/commits/6b439960b7a086f4e9ca2ce90d524352301984ec).

File: pkg/rpc/context.go
Comment (irfansharif) [inline pkg/rpc/context.go:1150]: [nit]

```go
if err := interceptor(req); err != nil {
  return err
}

// ...
```

File: pkg/rpc/context.go
Comment (irfansharif) [inline pkg/rpc/context.go:411]: Could we just set them to noops here, during validation/setting defaults, instead of down below?

**Review by irfansharif — COMMENTED** [review]
This all LGTM. 

I did manage to confuse myself a bit though: Contrasting it with what we were thinking earlier with persisting a "prevent startup file" (and with `crdb node decommission --force`) like we did in https://github.com/cockroachdb/cockroach/pull/54373, I realize the guarantees provided by that approach are a bit stronger (unless I'm missing something). Because there we persist the DECOMMISSIONED.txt file remotely _first_ before updating the liveness record, we could have the target node (the one being decommissioned) intentionally stop sending out PingRequests as part of that FinalizeDecommission API. I think it would have the same effect as this one in that all live connections are abandoned (except of course it would no longer be best effort). The target node would also be unable to restart because of that file, so that problem too goes away. 

As for when the target node is permanently downed (and --force is specified), then I think we'd need the safeguards we're introducing in this PR (stopping outgoing pings to the decomm'ed nodes, which we might as well do all the time). 

So I guess my question is, do we still need something like #54373? These failure modes do seem exceedingly rare, so I'm unsure.

**Review by tbg — COMMENTED** [review]
> Dismissed @knz from 3 discussions.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54936)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif and @knz)

___
*[pkg/kv/kvserver/node_liveness.go, line 243 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/54936#-MIOfMv6Ngi1sqOZcPiV-r10-243:-MITEcYc9BEFdNbIcxBp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/62b50cdb9864ada71e8bcccf1c712ad98c443e21/pkg/kv/kvserver/node_liveness.go#L243)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

s/option/options
</blockquote></details>

Done.
___
*[pkg/kv/kvserver/node_liveness.go, line 633 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/54936#-MIOfMv6Ngi1sqOZcPiV-r10-633:-MITEjz42MY_ncZ3Z8bo:b-gp9mm0) ([raw file](https://github.com/cockroachdb/cockroach/blob/62b50cdb9864ada71e8bcccf1c712ad98c443e21/pkg/kv/kvserver/node_liveness.go#L633)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Could move this comment to NodeLiveness.mu.onSelfLive, probably the first place to look.
</blockquote></details>

I referenced this comment from there.
___
*[pkg/kv/kvserver/node_liveness.go, line 1308 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/54936#-MIOfMv6Ngi1sqOZcPiV-r11-1308:-MITEvDzDWOCfzGxNxDP:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f77e99c0077f42999fd74653696dd75084ab80c/pkg/kv/kvserver/node_liveness.go#L1308)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

We have a helper, `newLivenessRec.Membership.Decommissioned()`.
</blockquote></details>

Done.
___
*[pkg/rpc/context.go, line 411 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/54936#-MIOfMv9Y3l3bH_fCoEq-r14-411:-MITEwGD9PfamCnvx82x:b93mb3y) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b4beb180bb077e1ee6f28433c788edfab50e2e3/pkg/rpc/context.go#L411)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Could we just set them to noops here, during validation/setting defaults, instead of down below?
</blockquote></details>

We could, but I find it easier to discover the defaults when they're applied at the single site of use. It's always kind of tedious to have to go chase down where the default is applied. It's mostly potato/potahto here, but I'll leave as is.
___
*[pkg/rpc/context.go, line 1150 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/54936#-MIOfMv9Y3l3bH_fCoEq-r14-1150:-MITGYGp7zU5LsE09IpK:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b4beb180bb077e1ee6f28433c788edfab50e2e3/pkg/rpc/context.go#L1150)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

[nit]

```go
if err := interceptor(req); err != nil {
  return err
}

// ...
```
</blockquote></details>

Done.

**Review by tbg — COMMENTED** [review]
We *can* extend the approach taken here to also write the prevented startup file and terminate the process when the RPC layer gets shut off from the cluster. However, I think we need to talk about the UX before we decide whether that's useful.
Users are generally expected to run their own supervisor, so they should be in charge of terminating the node. If we "disconnect" the node but the operator doesn't shut it down, it will definitely languish (i.e. SQL connections will hang, etc). I think that is "ok" for some value of "ok" - the operator supposedy took the node out of SQL load balancing before they initiated the decommissioning process.
We may want to consider a flag to `node decommission` in which the replicas are moved off, but we don't set the DECOMMISSIONED status yet. The idea would be that some users will prefer to move data off first, then shut down the node, and then set the DECOMMISSIONED status

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54936#-:-MIZQfbE7j5awjJ9XlYy:bdqfjzc)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif and @knz)

**Comment by tbg** [comment]
I prepended #55277 here and reworked `kvserver: add Nodeliveness.StartOptions.OnNodeDecommissionedCallback`. PTAL, should turn green now & be ready to go.

**Comment by tbg** [comment]
Friendly ping @irfansharif

**Review by irfansharif — APPROVED** [review]
:lgtm: I'll bookmark this PR for if/when we revisit decommissioning UX.

> We may want to consider a flag to node decommission in which the replicas are moved off, but we don't set the DECOMMISSIONED status yet. The idea would be that some users will prefer to move data off first, then shut down the node, and then set the DECOMMISSIONED status

I also like this idea, it removes another conceptual hurdle when having to explain the difference between a decommissioned node and dead one (and all our questions around decommissioning nodes in absentia go away).

> Reviewed 15 of 15 files at r15, 5 of 5 files at r16, 4 of 4 files at r17, 1 of 1 files at r18, 4 of 4 files at r19, 5 of 5 files at r20, 3 of 3 files at r21, 4 of 4 files at r22.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54936#-:-MJYIYs0C1ciqCl1wkVc:b-17rd9y)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @tbg)

___
*[pkg/rpc/context.go, line 383 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/54936#-MJYHR7I0c-CEDmh9Haj:-MJYHR7I0c-CEDmh9Hak:bjsawg8) ([raw file](https://github.com/cockroachdb/cockroach/blob/8552175a907be30c5a41982eedc838854fc4b378/pkg/rpc/context.go#L383)):*
> ```Go
> 	//
> 	// It can inject an error.
> 	OnHandlePing func(*PingRequest) error
> ```

[nit] Minor preference to name these On{Outgoing,Incoming}Ping instead.
___
*[pkg/rpc/heartbeat.go, line 56 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/54936#-MJYHh0a9UY72MY_OvAy:-MJYHh0a9UY72MY_OvAz:bs4q9op) ([raw file](https://github.com/cockroachdb/cockroach/blob/8552175a907be30c5a41982eedc838854fc4b378/pkg/rpc/heartbeat.go#L56)):*
> ```Go
> 	disableClusterNameVerification bool
> 
> 	onHandlePing func(*PingRequest) error // see ContextOptions.OnHandlePing
> ```

Ditto.

**Comment by tbg** [comment]
TFTR! I'll make the rename separately.

bors r=irfansharif

**Pull request merged**
